src/expand.c

```diff
@@ -0,0 +1,138 @@
+#include "vctrs.h"
+#include "type-data-frame.h"
+
+#include "decl/expand-decl.h"
+
+r_obj* ffi_vec_expand_grid(r_obj* xs,
+                           r_obj* ffi_vary,
+                           r_obj* ffi_name_repair,
+                           r_obj* frame) {
+  struct r_lazy error_call = { .x = syms.dot_error_call, .env = frame };
+
+  enum vctrs_expand_vary vary = parse_vary(ffi_vary);
+
+  struct name_repair_opts name_repair_opts = new_name_repair_opts(
+    ffi_name_repair,
+    lazy_args.dot_name_repair,
+    false,
+    error_call
+  );
+  KEEP(name_repair_opts.shelter);
+
+  r_obj* out = vec_expand_grid(xs, vary, &name_repair_opts, error_call);
+
+  FREE(1);
+  return out;
+}
+
+r_obj* vec_expand_grid(r_obj* xs,
+                       enum vctrs_expand_vary vary,
+                       const struct name_repair_opts* p_name_repair_opts,
+                       struct r_lazy error_call) {
+  vec_check_list(xs, vec_args.empty, error_call);
+
+  if (vec_any_missing(xs)) {
+    // Drop `NULL`s before any other checks
+    r_obj* complete = KEEP(vec_detect_complete(xs));
+    xs = vec_slice(xs, complete);
+    FREE(1);
+  }
+  KEEP(xs);
+
+  const r_ssize n = r_length(xs);
+
+  r_obj* out = KEEP(r_alloc_list(n));
+
+  r_obj* names = KEEP(vec_names2(xs));
+  if (!r_is_minimal_names(names)) {
+    r_abort_lazy_call(error_call, "Each element must be named.");
```

Comment:
> unpacking unnamed dfs

`tidyr::expand_grid()` but it feels like a weird behavior to me. It is kind of ambiguous to determine if you should unpack before or after the grid expansion (i.e. should the df be treated like 1 unit or as separate columns?). Treating it like one unit generally makes more sense, but it still seemed a little odd to me.

If it seems useful we can add it back in later. I figured it was better to be conservative for now since it is an easy thing to relax later if it feels useful